while-loops/while-loops.md

```diff
@@ -0,0 +1,257 @@
+
+Here, we should observe that at some point, `i` has a value of `6` , and the code asked "Is 6 <= 5? Is that true?" Because the answer is false, the code skips over the rest of the loop.
+
+Callout: While loops can accidentally run infinitely! If the conditional for the while loop never becomes false, then the while loop and the program keep going, and going, and going... until something forces it to stop (such as a human closing down the program), or our machine cannot run it anymore. We call these cases **infinite loops, ** and are always a bug.
+
+## Used While Problem-Solving
+
+Let's explore different examples that apply while loops.
+
+We can use a while loop even if we are decrementing a variable...
+
+``` python
+print("Preparing for blastoff in")
+
+countdown = 10
+
+while countdown > 0:
+    print(f"{countdown}...")
+    countdown -= 1
+
+print("Launch!")
+```
+
+We can use a while loop that uses a function to calculate something...
+
+``` python
+flower_basket = []
+
+while len(flower_basket) < 12:
+    flower_basket.append("flower")
+
+print("I have a very full flower basket with a dozen flowers:", flower_basket)
+```
+
+Callout: Note that this conditional uses `<` , _not_ `<=` , because we don't want to add another flower if the length of `flower_basket` is _equal_ to twelve. Watch out for an off-by-one error!
+
+We can use a while loop that only checks if a variable is truthy!
+
+``` python
+import random
+coin_is_tails = True
+
+while coin_is_tails:
+    print("The coin flip is tails!")
+    print("Will we get heads in the next coin flip?")
+    coin_is_tails = bool(random.randint(0, 1))
+
+print("We finally got heads.")
+```
+
+### Counter Variables
+
+Let's return to our original example:
+
+``` python
+i = 1
+while i <= 5:
+    print(f"The value of i is {i}")
+    i += 1
+
+print("We are outside the while loop! Wooo!")
+```
+
+A variable whose job it is to increment or decrement, specifically in a loop, is often called a _counter variable._
+
+This counter variable is:
+
+* used in the conditional expression for the while loop
+* must eventually be modified in the while loop
+* typically initialized before the loop
+
+In this case, `i` is the counter variable. It gets setup before the while loop begins, and then increments inside the while loop. When we forget to change our counter variable, we often get infinite loops and bugs!
+
+Callout: An alternative syntax is to use for loops with a range. The for loop syntax sets up and modifies something like a counter variable!
+
+### `break` and `continue`
+
+`break` is a keyword. When our code comes across `break` in a while loop, it will exit the entire while loop. The next line that runs is the line after the while loop.
+
+``` python
+def demonstrate_break():
+    i = 3
+
+    while i < 10:
+        if i == 7:
+            print("i is seven! Let's get out of here!")
+            break
+        print(f"Counting: i is {i}")
+        i += 1
+
+demonstrate_break()
+```
+
```

Comment:
I like the suggestion; I'm going to review this lesson visually before I make a call on adding this. Once I get a good visual on it, I'm going to assess if this output box pushes the lesson over into "this is extremely long and my brain is leaky" territory

One reason why this suggestion is in a medium-pri for me is because intro to `for` loops has them do ~20 loops, often with a loop table, so my hope is that this output is only reinforcement. But I def see the value in this output!